clientv3: make saved snapshot readable by user only

dlipovetsky opened the pull request:
Fixes #9976

dlipovetsky commented:
Note for reviewer:  Previously, the file was opened with [`os.Create`](https://golang.org/pkg/os/#Create), which opens the file for reading and writing (`os.O_RDWR`). Because the file does not need to be read, in this change it is opened for writing only (`os.O_WRONLY`).

hexfusion commented on clientv3/snapshot/v3_snapshot_test.go:
```diff
@@ -24,6 +24,8 @@ import (
 
+	"github.com/coreos/etcd/pkg/fileutil"
+
 	"github.com/coreos/etcd/clientv3"
```
extra newline

hexfusion commented on clientv3/snapshot/v3_snapshot.go:
```diff
@@ -102,7 +102,7 @@ func (s *v3Manager) Save(ctx context.Context, cfg clientv3.Config, dbPath string
-	f, err = os.Create(partpath)
+	f, err = os.OpenFile(partpath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, fileutil.PrivateFileMode)
 	if err != nil {
 		return fmt.Errorf("could not open %s (%v)", partpath, err)
```
If possible could you verify this works on Windows?

dlipovetsky replied:
Good point. I'll look into this.

The [`os.Chmod` docs](https://golang.org/pkg/os/#Chmod) indicates that on Windows, this change will not affect file permissions:

> On Windows, the mode must be non-zero but otherwise only the 0200 bit (owner writable) of mode is used; it controls whether the file's read-only attribute is set or cleared. attribute. The other bits are currently unused. Use mode 0400 for a read-only file and 0600 for a readable+writable file.

I'll run an experiment on Windows and update with the results.

hexfusion commented:
@dlipovetsky, thank you for the PR a few knits here but otherwise, this works as expected for me locally. As the previous permissions of the snapshot are something that could be assumed in various workflows we will need to document this possible breaking change well. Also as I noted in the review if we could verify the OpenFile usage work fine in Windows I think that would be smart as I am not sure our tests cover this.

I guess the final question is should etcdctl be in charge of this. While it should be up to the admin to provide a secure env I think that giving the user a little help and being proactive about something as important as a snapshot is a good thing and a step forward so I vote +1.

dlipovetsky commented:
I ran `etcdctl snapshot save` and `etcdctl snapshot restore` on Windows 2012 Server. Both work as expected:
```
C:\Users\Administrator\Desktop>etcdctl.exe snapshot save test.snapshot
{"level":"info","ts":1533346962.415343,"caller":"snapshot/v3_snapshot.go:109","msg":"created temporary db file","path":"test.snapshot.part"}
{"level":"info","ts":1533346962.4163485,"caller":"snapshot/v3_snapshot.go:120","msg":"fetching snapshot","endpoint":"127.0.0.1:2379"}
{"level":"info","ts":1533346962.4261172,"caller":"snapshot/v3_snapshot.go:133","msg":"fetched snapshot","endpoint":"127.0.0.1:2379","took":0.0097689}
{"level":"info","ts":1533346962.4261172,"caller":"snapshot/v3_snapshot.go:142","msg":"saved","path":"test.snapshot"}
Snapshot saved at test.snapshot
```
```
C:\Users\Administrator\Desktop>etcdctl.exe snapshot restore --data-dir test.etcd test.snapshot
{"level":"info","ts":1533347158.2669117,"caller":"snapshot/v3_snapshot.go:278","msg":"restoring snapshot","path":"test.snapshot","wal-dir":"test.etcd\\member\\w
al","data-dir":"test.etcd","snap-dir":"test.etcd\\member\\snap"}
{"level":"info","ts":1533347158.353891,"caller":"membership/cluster.go:343","msg":"added member","cluster-id":"cdf818194e3a8c32","local-member-id":"0","added-pe
er-id":"8e9e05c52164694d","added-peer-peer-urls":["http://localhost:2380"]}
{"level":"info","ts":1533347158.3782656,"caller":"wal/wal.go:252","msg":"closing WAL to release flock and retry directory renaming","from":"test.etcd\\member\\w
al.tmp","to":"test.etcd\\member\\wal"}
{"level":"info","ts":1533347158.4029384,"caller":"snapshot/v3_snapshot.go:291","msg":"restored snapshot","path":"test.snapshot","wal-dir":"test.etcd\\member\\wa
l","data-dir":"test.etcd","snap-dir":"test.etcd\\member\\snap"}
```
![image](https://user-images.githubusercontent.com/3445370/43671365-5832f4f2-974d-11e8-883a-7ad8f350cde4.png)

dlipovetsky commented:
@hexfusion I think my experiment demonstrates that this PR does not break `snapshot save` functionality on Windows. I also addressed your review comment. Once I hear back from you, and if there are no other review comments, I'll squash the commits. Thanks!

hexfusion commented:
@dlipovetsky as you may have seen a lot of effort is being put into the CNCF TOC meeting which is tomorrow. Please allow a few days for this PR to get squared away as I want to get additional feedback from the community. Thank you for the detailed Windows tests I am very happy we had no regression there.

dlipovetsky commented:
@hexfusion Thanks for your feedback! I'll keep an eye for an update.

gyuho commented:
Can you fix https://travis-ci.org/coreos/etcd/jobs/411980747#L607?

This should be safe, since we also use private file mode for WAL. And when snapshot file is fetched from remote storage, the user can always updates its file permission for their own use case.

dlipovetsky commented:
@gyuho Thanks for your feedback. The test failure should be be resolved when I squash the commits. I'll do that now.

hexfusion commented:
@dlipovetsky thank you for your contribution, I hope this is the first of many.

gyuho commented:
@dlipovetsky Could you also add change log here https://github.com/coreos/etcd/blob/master/CHANGELOG-3.4.md#client-v3, so people know who worked on it?

dlipovetsky commented:
@hexfusion Thanks for guiding me along!
@gyuho Will submit a PR for the changelog, thanks!
